lightning/src/offers/invoice_request.rs

```diff
@@ -0,0 +1,1259 @@
+use bitcoin::secp256k1::{Message, PublicKey};
+use bitcoin::secp256k1::schnorr::Signature;
+use core::convert::TryFrom;
+use crate::io;
+use crate::ln::features::InvoiceRequestFeatures;
+use crate::ln::msgs::DecodeError;
+use crate::offers::merkle::{SignError, SignatureTlvStream, SignatureTlvStreamRef, self};
+use crate::offers::offer::{Offer, OfferContents, OfferTlvStream, OfferTlvStreamRef};
+use crate::offers::parse::{ParseError, ParsedMessage, SemanticError};
+use crate::offers::payer::{PayerContents, PayerTlvStream, PayerTlvStreamRef};
+use crate::util::ser::{HighZeroBytesDroppedBigSize, SeekReadable, WithoutLength, Writeable, Writer};
+use crate::util::string::PrintableString;
+
+use crate::prelude::*;
+
+const SIGNATURE_TAG: &'static str = concat!("lightning", "invoice_request", "signature");
+
+/// Builds an [`InvoiceRequest`] from an [`Offer`] for the "offer to be paid" flow.
+///
+/// See [module-level documentation] for usage.
+///
+/// [module-level documentation]: self
+pub struct InvoiceRequestBuilder<'a> {
+	offer: &'a Offer,
+	invoice_request: InvoiceRequestContents,
+}
+
+impl<'a> InvoiceRequestBuilder<'a> {
+	pub(super) fn new(offer: &'a Offer, metadata: Vec<u8>, payer_id: PublicKey) -> Self {
+		Self {
+			offer,
+			invoice_request: InvoiceRequestContents {
+				payer: PayerContents(metadata), offer: offer.contents.clone(), chain: None,
+				amount_msats: None, features: InvoiceRequestFeatures::empty(), quantity: None,
+				payer_id, payer_note: None,
+			},
+		}
+	}
+
+	/// Sets the [`InvoiceRequest::chain`] of the given [`Network`] for paying an invoice. If not
+	/// called, [`Network::Bitcoin`] is assumed. Must be supported by the offer.
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn chain(mut self, network: Network) -> Result<Self, SemanticError> {
+		let chain = ChainHash::using_genesis_block(network);
+		if !self.offer.supports_chain(chain) {
+			return Err(SemanticError::UnsupportedChain);
+		}
+
+		self.invoice_request.chain = Some(chain);
+		Ok(self)
+	}
+
+	#[cfg(test)]
+	pub fn chain_unchecked(mut self, network: Network) -> Self {
+		let chain = ChainHash::using_genesis_block(network);
+		self.invoice_request.chain = Some(chain);
+		self
+	}
+
+	/// Sets the [`InvoiceRequest::amount_msats`] for paying an invoice. Must be at least the base
+	/// invoice amount (i.e., [`Offer::amount`] times [`quantity`]).
+	///
+	/// Successive calls to this method will override the previous setting.
+	///
+	/// [`quantity`]: Self::quantity
+	pub fn amount_msats(mut self, amount_msats: u64) -> Result<Self, SemanticError> {
+		self.invoice_request.offer.check_amount_msats(
+			Some(amount_msats), self.invoice_request.quantity
+		)?;
+		self.invoice_request.amount_msats = Some(amount_msats);
+		Ok(self)
+	}
+
+	#[cfg(test)]
+	pub fn amount_msats_unchecked(mut self, amount_msats: u64) -> Self {
+		self.invoice_request.amount_msats = Some(amount_msats);
+		self
+	}
```

Comment:
Reviewing-wise I think it would be nicer if they were at least grouped together under one `cfg(test) impl InvoiceRequest { .. }`, since e.g. I want to review the real code 10x and the test code once, but not a blocker at all